src/software/new_geom/util/closest_point.cpp

```diff
@@ -15,3 +15,103 @@ Point closestPointOnLine(const Line &l, const Point &p)
 {
     return closestPointOnLine(p, l);
 }
+
+Point closestPointOnSegment(const Point &p, const Segment &segment)
+{
+    // if one of the end-points is extremely close to the centre point
+    // then return 0.0
+    if ((segment.getEnd() - p).lengthSquared() < FIXED_EPSILON * FIXED_EPSILON)
+    {
+        return segment.getEnd();
+    }
+
+    if ((segment.getStart() - p).lengthSquared() < FIXED_EPSILON * FIXED_EPSILON)
+    {
+        return segment.getStart();
+    }
+
+    // take care of 0 length segments
+    if ((segment.getEnd() - segment.getStart()).lengthSquared() <
+        FIXED_EPSILON * FIXED_EPSILON)
+    {
+        return segment.getStart();
+    }
+
+    // find point C
+    // which is the projection onto the line
+    double lenseg = (segment.getEnd() - segment.getStart()).dot(p - segment.getStart()) /
+                    (segment.getEnd() - segment.getStart()).length();
+    Point C =
+        segment.getStart() + lenseg * (segment.getEnd() - segment.getStart()).normalize();
+
+    // check if C is in the line seg range
+    double AC     = (segment.getStart() - C).lengthSquared();
+    double BC     = (segment.getEnd() - C).lengthSquared();
+    double AB     = (segment.getStart() - segment.getEnd()).lengthSquared();
+    bool in_range = AC <= AB && BC <= AB;
+
+    // if so return C
+    if (in_range)
+    {
+        return C;
+    }
+    double lenA = (p - segment.getStart()).length();
+    double lenB = (p - segment.getEnd()).length();
+
+    // otherwise return closest end of line-seg
+    if (lenA < lenB)
+    {
+        return segment.getStart();
+    }
+    return segment.getEnd();
+}
+
+Point closestPointOnSegment(const Segment &segment, const Point &p)
+{
```

Comment:
this function should call the other function with reversed arguments

Reply:
Done.